apm-agent-plugins/apm-profiling-plugin/src/test/java/co/elastic/apm/agent/profiler/RecoverSamplingProfilerTest.java

```diff
@@ -0,0 +1,182 @@
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.profiler;
+
+import co.elastic.apm.agent.MockReporter;
+import co.elastic.apm.agent.MockTracer;
+import co.elastic.apm.agent.configuration.SpyConfiguration;
+import co.elastic.apm.agent.configuration.converter.TimeDuration;
+import co.elastic.apm.agent.impl.ElasticApmTracer;
+import co.elastic.apm.agent.impl.Scope;
+import co.elastic.apm.agent.impl.transaction.Span;
+import co.elastic.apm.agent.impl.transaction.Transaction;
+import co.elastic.apm.agent.matcher.WildcardMatcher;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.condition.DisabledOnOs;
+import org.junit.jupiter.api.condition.OS;
+import org.stagemonitor.configuration.ConfigurationRegistry;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.awaitility.Awaitility.await;
+import static org.mockito.Mockito.when;
+
+// async-profiler doesn't work on Windows
+@DisabledOnOs(OS.WINDOWS)
+class RecoverSamplingProfilerTest {
+
+    private MockReporter reporter;
+    private ElasticApmTracer tracer;
+    private SamplingProfiler profiler;
+    private ProfilingConfiguration profilingConfig;
+
+    @BeforeEach
+    void setUp() {
+        reporter = new MockReporter();
+        ConfigurationRegistry config = SpyConfiguration.createSpyConfig();
+        profilingConfig = config.getConfig(ProfilingConfiguration.class);
+        when(profilingConfig.getIncludedClasses()).thenReturn(List.of(WildcardMatcher.valueOf(getClass().getName())));
+        when(profilingConfig.isProfilingEnabled()).thenReturn(true);
+        when(profilingConfig.getProfilingDuration()).thenReturn(TimeDuration.of("500ms"));
+        when(profilingConfig.getProfilingInterval()).thenReturn(TimeDuration.of("500ms"));
+        when(profilingConfig.getSamplingInterval()).thenReturn(TimeDuration.of("5ms"));
+        tracer = MockTracer.createRealTracer(reporter, config);
+        profiler = tracer.getLifecycleListener(ProfilingFactory.class).getProfiler();
+        // ensure profiler is initialized
+        await()
+            .pollDelay(10, TimeUnit.MILLISECONDS)
+            .timeout(6000, TimeUnit.MILLISECONDS)
+            .until(() -> profiler.getProfilingSessions() > 1);
+    }
+
+    @AfterEach
+    void tearDown() throws Exception {
+        tracer.stop();
+    }
+
+    @Test
+    void testAll() throws Exception {
```

Comment:
What's the value of testing the different options?

Reply:
Only sanity check that enabling discrete parts of the recovery logic doesn't break anything. Again - not to be eventually merged.